### PR TITLE
Auth password config rework

### PR DESCRIPTION
**Pull request by chrzaszcz**
The goal of this PR is to rework the `auth.password` config section to:

- Use **maps**. The resulting format is also a map now. The requirement of not having the `hash` option if the password format is `plain` is removed as it was undocumented, inconsistent with no such check for `scram_iterations` and prevented the defaults from being set. It seems enough to inform the user that the `plain` format just ignores these options.
- Use **defaults** - only the `hash` option is left without a default. Putting the default list of hashes there might lead to suboptimal code, as `lists:member` would be called on each call to `listmech` for each method. This can be changed when we rework the `cyrsasl` implementation, e.g. by precomputing effective filtered SASL mechanisms per host type, but it falls out of scope of this PR. 
- Contain the `scram_iterations` option, which is password-related, but was outside of this section.

### CODE REVIEW

**Comment by mongoose-im** [comment]
[small_tests_23](https://circleci.com/gh/esl/MongooseIM/95925) / small_tests / 6459bddb12e4a49502ee96686cc5eb8cea313d33
Reports [root](http://esl.github.io/circleci-mim-results/s3_reports.html?prefix=PR/3463/95925/small_tests.23.1) / [small](https://circleci-mim-results.s3.eu-central-1.amazonaws.com/PR/3463/95925/small_tests.23.1/small/ct_run.mongooseim@localhost.2021-12-20_11.36.22/lib.mongooseim.logs/run.2021-12-20_11.36.22/suite.log.html)

**Comment by mongoose-im** [comment]
[small_tests_24](https://circleci.com/gh/esl/MongooseIM/95966) / small_tests / 589d0a6a31a8485bc85350d243e48cf03025f765
Reports [root](http://esl.github.io/circleci-mim-results/s3_reports.html?prefix=PR/3463/95966/small_tests.24.0.5) / [small](https://circleci-mim-results.s3.eu-central-1.amazonaws.com/PR/3463/95966/small_tests.24.0.5/small/ct_run.mongooseim@localhost.2021-12-20_12.20.50/lib.mongooseim.logs/run.2021-12-20_12.20.51/suite.log.html)

File: src/mongoose_scram.erl [hunk 1]
```diff
@@ -74,18 +70,16 @@ enabled(HostType, cyrsasl_scram_sha512_plus) -> is_password_format_allowed(HostT
 enabled(_HostType, _Mechanism) -> false.
 
 is_password_format_allowed(HostType, Sha) ->
-    case mongoose_config:get_opt([{auth, HostType}, password_format], scram) of
-        plain -> true;
-        scram -> true;
-        {scram, ConfiguredSha} -> lists:member(Sha, ConfiguredSha)
+    case mongoose_config:get_opt([{auth, HostType}, password]) of
+        #{format := scram, hash := ConfiguredSha} -> lists:member(Sha, ConfiguredSha);
+        #{format := _PlainOrScram} -> true
```
Comment (NelsonVides) [inline src/mongoose_scram.erl:75]: Nice, that's very nice, good optimisation against the common case, less `lists:member/2` 👏🏽

**Review by NelsonVides — APPROVED** [review]
Neat 👌🏽

**Pull request merged**